torch/distributed/fsdp/_fully_shard/_fsdp_common.py

```diff
@@ -179,3 +179,25 @@ def _cast_fp_tensor(dtype: torch.dtype, x: torch.Tensor) -> torch.Tensor:
     ):
         return x
     return x.to(dtype)
+
+
+class _Allocator(Protocol):
+    def __call__(
+        self, size: tuple[int, ...], *, dtype: torch.dtype, device: torch.device
+    ) -> torch.Tensor:
+        ...
+
+
+_fsdp_comm_allocator: Optional[_Allocator] = None
+
+
+def _set_fsdp_comm_allocator(allocator: _Allocator):
+    global _fsdp_comm_allocator
+    _fsdp_comm_allocator = allocator
+
+
+def _get_fsdp_comm_allocator() -> _Allocator:
+    if _fsdp_comm_allocator is not None:
+        return _fsdp_comm_allocator
+    else:
+        return torch.empty
```

Comment:
naming is hard 😅 

if you make these global, I am okay with that

that might be better than passing as args into `fully_shard` since then user is expected to pass the same allocator to all calls of `fully_shard` (or I think it does not make too much sense)

Reply:
> that might be better than passing as args into fully_shard since then user is expected to pass the same allocator to all calls of fully_shard (or I think it does not make too much sense)

Makes sense. I made it global mainly because some allocations are performed in custom ops created for tracing, and I don't want to mess them up.

Hmm I think we also need to expose the process group on which the collective will be performed 🤔

Reply:
Hmm right now it seems like the allocator doesn't know when the memory is not needed? My understanding is tensor doesn't have callback mechanism to tell allocator so, so wonder how GC is supposed to work?

Reply:
@xunnanxu you can pass your memory deallocation logic to `at::from_blob` ([example](https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/SymmetricMemory.cpp#L184-L189)). Do you think this would work?

Reply:
@yifuwang Got it. I got slightly confused by the example as the allocation is done in python but this makes sense.